// cors-bypass/lib/Adapter/Adapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var constants_1 = require("../constants");
var utils_1 = require("../utils");
var Adapter = /** @class */ (function () {
    function Adapter() {
        var _this = this;
        this.rx = new BroadcastChannel(constants_1.CHANNEL_RX);
        this.tx = new BroadcastChannel(constants_1.CHANNEL_TX);
        this.onMessage = function (_a) {
            var data = _a.data;
            if (!utils_1.isValidEvent(data))
                return;
            var event = data;
            _this.rx.postMessage(event);
        };
        this.postMessage = function (_a) {
            var data = _a.data;
            var event = utils_1.makeValidEvent(data);
            window.parent && window.parent.postMessage(event, '*');
        };
        window.addEventListener('message', this.onMessage);
        this.tx.addEventListener('message', this.postMessage);
    }
    Adapter.prototype.dispose = function () {
        window.removeEventListener('message', this.onMessage);
        this.tx.close();
    };
    return Adapter;
}());
exports.Adapter = Adapter;
