// cors-bypass/lib/Server/ConnectedClient/ConnectedClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Handler_1 = require("../../Handler");
var ConnectedClient = /** @class */ (function () {
    function ConnectedClient(server, id) {
        var _this = this;
        this.server = server;
        this.id = id;
        this.handlers = Handler_1.handlers.map(function (Handler) { return new Handler(_this.server, _this); });
    }
    ConnectedClient.prototype.handleEvent = function (topic, payload) {
        if (/constructor|dispose|prototype/.test(String(topic)))
            return null;
        for (var _i = 0, _a = this.handlers; _i < _a.length; _i++) {
            var handler = _a[_i];
            if (topic in handler) {
                var method = handler[topic];
                if (typeof method === 'function') {
                    var returnValue = method.bind(handler)(payload);
                    return { handler: handler, method: method, returnValue: returnValue };
                }
            }
        }
        return null;
    };
    ConnectedClient.prototype.emit = function (topic, data) {
        return this.server.emit(topic, data, this.id);
    };
    ConnectedClient.prototype.dispose = function () {
        this.handlers.forEach(function (handler) { return handler.dispose && handler.dispose(); });
        this.server.clients.delete(this.id);
    };
    return ConnectedClient;
}());
exports.ConnectedClient = ConnectedClient;
